// eslint-plugin-fsd/lib/utils/patterns.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.anyHtmlElementObjectCollection = exports.anyHtmlElementsSearchMethod = exports.anyListenersMethods = exports.anyClipboardEvent = exports.anyDragEvent = exports.anyMouseEvent = exports.anyKeyboardEvent = exports.anyFormEvent = exports.anyWindowEvent = void 0;
const anyWindowEvent = /^(?:on(?:afterprint|beforeprint|beforeunload|error|hashchange|load|message|offline|online|pagehide|pageshow|popstate|resize|storage|unload))&/u;
exports.anyWindowEvent = anyWindowEvent;
const anyFormEvent = /^(?:on(?:blur|change|contextmenu|focus|input|invalid|reset|search|select|submit))$/u;
exports.anyFormEvent = anyFormEvent;
const anyKeyboardEvent = /^(?:on(?:keydown|keypress|keyup))$/u;
exports.anyKeyboardEvent = anyKeyboardEvent;
const anyMouseEvent = /^(?:on(?:click|dblclick|mousedown|mousemove|mouseout|mouseover|mouseup|mousewheel|wheel))$/u;
exports.anyMouseEvent = anyMouseEvent;
const anyDragEvent = /^(?:on(?:drag|dragend|dragenter|dragleave|dragover|dragstart|drop|scroll))$/u;
exports.anyDragEvent = anyDragEvent;
const anyClipboardEvent = /^(?:on(?:copy|cut|paste))$/u;
exports.anyClipboardEvent = anyClipboardEvent;
const anyListenersMethods = /^(?:(?:add|remove)EventListener)$/u;
exports.anyListenersMethods = anyListenersMethods;
const anyHtmlElementsSearchMethod = /^(?:getElement(?:ById|sByTagName|sByClassName)|querySelector(?:All)?)$/u;
exports.anyHtmlElementsSearchMethod = anyHtmlElementsSearchMethod;
const anyHtmlElementObjectCollection = /^(?:anchors|body|documentElement|embeds|forms|head|images|links|scripts|title)$/;
exports.anyHtmlElementObjectCollection = anyHtmlElementObjectCollection;
