// eslint-plugin-fsd/lib/rules/jq-use-js-prefix-in-selector.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.errorMessages = exports.rule = exports.name = void 0;

var _experimentalUtils = require("@typescript-eslint/experimental-utils");

require("core-js");

//
// ─── HELPER FUNCTIONS ───────────────────────────────────────────────────────────
//

/**
 * Check if provided selector adheres to rule:
 * All classes that are used to select dom elements - should start with "js-" prefix
 * @param selector jquery selector string
 */
function isCorrectSelector(selector = '') {
  // if selector is an html element string
  if (selector.startsWith('<')) return true;
  const parts = selector.split(',').flatMap(s => s.split(/\s+/));
  const allSelectorsAreCorrect = parts.every(sel => {
    // if selector is a class - it should be prefixed with 'js-'
    if (sel.startsWith('.') && !sel.startsWith('.js-')) return false;
    return true;
  });
  return allSelectorsAreCorrect;
} //
// ─── RULE DECLARATIONS ──────────────────────────────────────────────────────────
//


const name = 'jq-use-js-prefix-in-selector';
exports.name = name;

const createRule = _experimentalUtils.ESLintUtils.RuleCreator(() => `https://github.com/fullstack-development/front-end-best-practices/blob/master/JS/jQuery.md`);

const errorMessages = {
  useJsPrefix: 'All classes that are used to search the DOM must begin with "js-" prefix'
};
exports.errorMessages = errorMessages;
const meta = {
  type: 'suggestion',
  docs: {
    category: 'Best Practices',
    description: 'All classes that are used to search the DOM must begin with "js-" prefix',
    recommended: false
  },
  messages: errorMessages,
  schema: []
};
const jqDOMSelector = 'MemberExpression > CallExpression > Identifier.callee[name=/\\$|(jQuery)/]';
const rule = createRule({
  name,
  meta,
  defaultOptions: [],

  create(context) {
    return {
      [jqDOMSelector]: function checkDOMSelector(node) {
        const callExpression = node.parent;
        const firstArg = callExpression.arguments[0];

        if (firstArg.type === _experimentalUtils.AST_NODE_TYPES.Literal) {
          if (!firstArg.value) return;
          const selector = firstArg.value.toString();

          if (!isCorrectSelector(selector)) {
            context.report({
              node: callExpression,
              messageId: 'useJsPrefix'
            });
          }
        }
      }
    };
  }

});
exports.rule = rule;
