// eslint-plugin-fsd/lib/rules/no-function-declaration-in-event-listener.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.errorMessages = exports.rule = exports.name = void 0;

var _experimentalUtils = require("@typescript-eslint/experimental-utils");

var _astUtils = require("eslint/lib/rules/utils/ast-utils");

//
// ─── RULE DECLARATIONS ──────────────────────────────────────────────────────────
//
const name = 'no-function-declaration-in-event-listener';
exports.name = name;

const createRule = _experimentalUtils.ESLintUtils.RuleCreator(() => 'https://github.com/fullstack-development/front-end-best-practices/blob/master/JS/goodPractice.md');

const errorMessages = {
  functionForEventListener: 'Move event handlers to separate functions'
};
exports.errorMessages = errorMessages;
const meta = {
  type: 'suggestion',
  docs: {
    category: 'Best Practices',
    description: 'Move event handlers to separate functions',
    recommended: false
  },
  messages: errorMessages,
  schema: []
};
const addEventListenerSelector = 'CallExpression > MemberExpression > Identifier.property[name=/addEventListener|on/]';
const rule = createRule({
  name,
  meta,
  defaultOptions: [],

  create(context) {
    return {
      [addEventListenerSelector]: function checkEventListener(node) {
        const memberExpression = node.parent;
        const callExpression = memberExpression.parent;
        const secondArg = callExpression.arguments[1];

        if ((0, _astUtils.isFunction)(secondArg)) {
          context.report({
            node: secondArg,
            messageId: 'functionForEventListener'
          });
        }
      }
    };
  }

});
exports.rule = rule;
