// eslint-plugin-fsd/lib/rules/no-heavy-constructor.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _typeguards = require("../utils/typeguards");

/**
 * @fileoverview Higher order functions name prefix
 * @author Leonid Baryshnikov
 */
//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------
const rule = {
  meta: {
    type: 'layout',
    docs: {
      description: 'No heavy constructor',
      category: 'Stylistic Issues',
      recommended: false,
      url: 'https://github.com/timon-and-pumbaa/eslint-plugin-fsd/blob/master/docs/rules/no-heavy-constructor.md'
    },
    fixable: undefined,
    schema: [],
    messages: {
      heavyConstructor: "Constructor shouldn't search the DOM and define handlers"
    }
  },

  create(context) {
    //----------------------------------------------------------------------
    // Helpers
    //----------------------------------------------------------------------
    const reportIdentifier = identifier => {
      context.report({
        node: identifier,
        messageId: 'heavyConstructor'
      });
    };

    const checkForEventAttributes = identifier => {
      if ((0, _typeguards.isHtmlEventIdentifier)(identifier)) {
        reportIdentifier(identifier);
      }
    };

    const checkForListenersMethods = identifier => {
      if ((0, _typeguards.isListenerMethodIdentifier)(identifier)) {
        reportIdentifier(identifier);
      }
    };

    const checkForSearchMethodsAndCollections = identifier => {
      if ((0, _typeguards.isHtmlElementSearchMethodIdentifier)(identifier) || (0, _typeguards.isHtmlElementsObjectCollectionIdentifier)(identifier)) {
        reportIdentifier(identifier);
      }
    }; //----------------------------------------------------------------------
    // Public
    //----------------------------------------------------------------------


    const constructor = 'ClassDeclaration MethodDefinition[key.name="constructor"]';
    const identifier = 'Identifier.property[name]';
    return {
      [`${constructor} MemberExpression.left ${identifier}`]: checkForEventAttributes,
      [`${constructor} MemberExpression.callee ${identifier}`]: checkForListenersMethods,
      [`${constructor} MemberExpression ${identifier}`]: checkForSearchMethodsAndCollections
    };
  }

};
var _default = rule;
exports.default = _default;
