// eslint-plugin-fsd/lib/rules/split-conditionals.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.errorMessages = exports.rule = exports.name = void 0;

var _experimentalUtils = require("@typescript-eslint/experimental-utils");

//
// ─── HELPER FUNCTIONS ───────────────────────────────────────────────────────────
//
function hasMoreThen1Condition(node) {
  return node.left.type === _experimentalUtils.AST_NODE_TYPES.LogicalExpression || node.right.type === _experimentalUtils.AST_NODE_TYPES.LogicalExpression;
} //
// ─── RULE DECLARATIONS ──────────────────────────────────────────────────────────
//


const name = 'split-conditionals';
exports.name = name;

const createRule = _experimentalUtils.ESLintUtils.RuleCreator(() => 'https://github.com/fullstack-development/front-end-best-practices/blob/master/JS/goodPractice.md');

const errorMessages = {
  tooManyConditions: 'All checks containing more than one condition must be separated'
};
exports.errorMessages = errorMessages;
const meta = {
  type: 'suggestion',
  docs: {
    category: 'Best Practices',
    description: 'All checks containing more than one condition must be separated',
    recommended: false
  },
  messages: errorMessages,
  schema: []
};
const rule = createRule({
  name,
  meta,
  defaultOptions: [],

  create(context) {
    function checkLogicalExpression(node) {
      if (hasMoreThen1Condition(node)) {
        context.report({
          node,
          messageId: 'tooManyConditions'
        });
      }
    }

    function checkRule(node) {
      const testCondition = node.test;
      if (!testCondition) return;
      if (testCondition.type !== _experimentalUtils.AST_NODE_TYPES.LogicalExpression) return;
      checkLogicalExpression(testCondition);
    }

    return {
      IfStatement: checkRule,
      WhileStatement: checkRule,
      DoWhileStatement: checkRule,
      ForStatement: checkRule
    };
  }

});
exports.rule = rule;
